### DISCUSSION TITLE
can fluidsynth initialize the COM library in the context of the synthesis thread instead of the caller's ?

### DISCUSSION THREAD

**Discussion by pedrolcl** [FluidSynth/fluidsynth#833] [General]
The wasapi driver included in fluidsynth 2.2.0 requests MTA COM thread initialization in the running thread with `CoInitializeEx(NULL, COINIT_MULTITHREADED)`, which [may fail](https://docs.microsoft.com/en-us/windows/win32/api/combaseapi/nf-combaseapi-coinitializeex#return-value) if the thread has been initialized already in STA model:
~~~
fluidsynth: error: wasapi: cannot initialize COM. 0x80010106
~~~
The error code 0x80010106 means RPC_E_CHANGED_MODE which is defined in WinError.h:
~~~
#define RPC_E_CHANGED_MODE               _HRESULT_TYPEDEF_(0x80010106L)
~~~
See also this [QSynth issue](https://github.com/rncbc/qsynth/issues/38).

It is understandable that the drivers require some special threading conditions. It already happens with the realtime priority of the Unix drivers, but ...

**Please, can fluidsynth initialize the COM library in the context of the synthesis thread instead of the caller's?**

**Comment by derselbst** [comment]
Moving the `CoInitializeEx` to the synth worker thread would allow getting rid of init in `new_fluid_audio_driver2`. But there is a second init that we call whenever a settings object is created. I see no other chance than to perform all those COM calls in a worker. Which would be a bit ugly as well.

 Is there a way to test whether COM has been initialized already? @chirs241097 Do you have any other idea on this?

**Reply by pedrolcl** [reply]
You are right, there is a second call to `CoInitializeEx` in `fluid_wasapi_foreach_device`. Here we need another background thread.

When `CoInitializeEx` fails with the code `RPC_E_CHANGED_MODE`, this means that the thread was already initialized with a different model. There is no other way to test it, AFAIK.

In QSynth, the sequence when selecting the new wasapi driver is:

- At the beginning, the UI thread is initialized in COM STA model by Qt. Every fluidsynth API function is called from this thread.
- `fluid_settings_foreach` is called, which indirectly calls `fluid_wasapi_foreach_device`, and tries `CoInitializeEx`, which fails with the ` RPC_E_CHANGED_MODE` error code. The consequence is that the list of audio devices in the settings object becomes empty at this point, but then `CoUninitialize` is called, and the former thread attribute is cleared.
- `new_fluid_audio_driver2` is called, and `CoInitializeEx` succeeds this time. The wasapi audio driver works, using the default device (you can't select a different one).
- At this point, the UI thread has changed to MTA model, and the consequences for the GUI are:
   - drag and drop no longer works 
   - the Open/Save dialogs do not work anymore, and the window freezes when you try.

Solutions? All clients should call the impacted fluidsynth functions from background threads. This is what I've done in Drumstick already. We should tell every project using the fluidsynth library to do the same, if they plan to use the new wasapi driver. This is the only driver with this inconvenience. 

Or we can avoid the inconvenience for the clients, centralizing in the wasapi driver the creation of background threads when they need to specify thread settings. As an example, most Unix drivers need realtime priority threads (with the excepion of jack) and each driver creates its threads requesting this attribute without changing anything in the parent process.

**Reply by derselbst** [reply]
But in `fluid_wasapi_foreach_device` we do not share any COM interfaces with other threads. So in this case we could simply ignore a `RPC_E_CHANGED_MODE` and silently continue? (+avoiding the `CoUninitialize` call at the end)

In `new_fluid_audio_driver2` the only COM object we share with another thread is the one created by `IMMDevice_Activate`. Here, I agree that this initialization should be moved to the synth thread.

**Reply by pedrolcl** [reply]
> But in fluid_wasapi_foreach_device we do not share any COM interfaces with other threads. So in this case we could simply ignore a RPC_E_CHANGED_MODE and silently continue? (+avoiding the CoUninitialize call at the end)

And ignore the failure in the device enumeration? When the init fails, you get an empty list of available devices in the settings object, so only the "default" device would be offered to the user. If this is acceptable, then we can avoid calling `fluid_wasapi_foreach_device` entirely in `fluid_wasapi_audio_driver_settings`.

**Reply by derselbst** [reply]
No, from my understanding `RPC_E_CHANGED_MODE` indicates that COM is already inited and stays inited with the previous thread model. So it should be safe to continue with the device enumeration as COM stays inited. Currently, as I understand, the `FAILED(ret)` is responsible for cancelling the device enumeration. So I suggest to change it to

```c
int com_was_inited = FALSE;
ret = CoInitializeEx(NULL, COINIT_MULTITHREADED);
if(ret == S_FALSE)
{
  goto cleanup;
}
else if(ret == RPC_E_CHANGED_MODE)
{
  com_was_inited = TRUE;
}

// ...dev enumeration...

if(!com_was_inited)
{
  CoUninitialize();
}
```

**Reply by pedrolcl** [reply]
The [documentation](https://docs.microsoft.com/en-us/cpp/parallel/concrt/walkthrough-using-the-concurrency-runtime-in-a-com-enabled-application?view=msvc-160#managing-the-lifetime-of-the-com-library) says:

> For each call to CoInitializeEx, a thread must also call CoUninitialize. In other words, calls to CoInitializeEx and CoUninitialize must be balanced.

In another [remark](https://docs.microsoft.com/en-us/windows/win32/api/combaseapi/nf-combaseapi-couninitialize#remarks) it says:

> A thread must call CoUninitialize once for each successful call it has made to the CoInitialize or CoInitializeEx function, including any call that returns S_FALSE. 

So, my understanding is that only if `SUCCEEDED(ret)` we should call `CoUninitialize`.

Anyway, IMHO the problem is that a library should not mess with the threads that has not created, like the ones that belong to the client process.

**Reply by derselbst** [reply]
> So, my understanding is that only if SUCCEEDED(ret) we should call CoUninitialize.

Yes, that's also my understanding and goes hand in hand with the `CCoInitializer` example class listed in your first doc link. My code example does basically the same. This is why I think we do not need another worker thread in `fluid_wasapi_foreach_device`.

My problem with worker threads is that you'll never know when they are done. You would probably wait for some quit event for a certain time. But who can tell the right value for such a timeout? If the timeout is too short, people like to call `TerminateThread()` while buying all the problems it may or may not bring. So whenever possible I would like to avoid this. And in this case it seems possible to avoid it at all.

**Reply by pedrolcl** [reply]
OK. I've tested the strategy discussed here, and it seems to work. You have a PR.

**Comment by chirs241097** [comment]
Well, I saw the latest commit already resolved the issue. And I skimmed through the changes, they do make a lot of sense.
Messing with the caller thread's COM concurrency model is definitely unacceptable. Thanks a lot for the fix!

(Sorry for the late response. Had to take a couple of weeks off from github for personal reasons.)

**Reply by pedrolcl** [reply]
Thanks to you too for implementing the driver!
